### ISSUE TITLE
OutOfMemoryError thown when making multiple silentAuthAsync calls | Stress Test

### ISSUE THREAD

**Issue by heidijinxujia** [AzureAD/azure-activedirectory-library-for-android#824]
*Stress Test Details*
Open 10 threads calling acquireTokenSilient at the same time for 8 hours with empty cache. 

*Problem*
After one hour running, got the OutOfMemory error. I could not use the semaphore here, because the mHandler is queuing all the msg which prevent the calls to hit the callback after getting the auth result. Without hitting the callback, the permit in the semaphore cannot be released. I've tried use a bounded queue for each thread, but got the same OutOfMemory.

```
Process: com.microsoft.aad.adal.userappwithbroker, PID: 1971
java.lang.OutOfMemoryError: Failed to allocate a 1054 byte allocation with 4194304 free bytes and 4MB until OOM; failed due to fragmentation (required continguous free 131072 bytes for a new buffer where largest contiguous free 69632 bytes)
    at java.lang.StringFactory.newStringFromChars(Native Method)
    at java.lang.StringFactory.newStringFromChars(StringFactory.java:218)
    at java.lang.StringFactory.newStringFromBytes(StringFactory.java:203)
    at java.lang.StringFactory.newStringFromBytes(StringFactory.java:63)
    at com.microsoft.aad.adal.IdToken.parseJWT(IdToken.java:136)
    at com.microsoft.aad.adal.IdToken.<init>(IdToken.java:66)
    at com.microsoft.aad.adal.APIEvent.setIdToken(APIEvent.java:123)
    at com.microsoft.aad.adal.AcquireTokenRequest.performAcquireTokenRequest(AcquireTokenRequest.java:277)
    at com.microsoft.aad.adal.AcquireTokenRequest.access$200(AcquireTokenRequest.java:46)
    at com.microsoft.aad.adal.AcquireTokenRequest$1.run(AcquireTokenRequest.java:109)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
    at java.lang.Thread.run(Thread.java:818)
```

*Code*
```
private void callEmptyCacheStressTest(final String User_UPN) throws InterruptedException{
	final int MAX_AVAILABLE = 10;
	final int TIME_LIMIT = 8 * 60 * 60 * 1000;
	final long startTime = System.currentTimeMillis();

	//clear the cache
	mAuthContext.getCache().removeAll();

	BlockingQueue<Runnable> blockingQueue = new ArrayBlockingQueue<Runnable>(10);
	RejectedExecutionHandler rejectedExecutionHandler = new ThreadPoolExecutor.CallerRunsPolicy();
	ExecutorService executor = new ThreadPoolExecutor(1, MAX_AVAILABLE, 0L, TimeUnit.MILLISECONDS, blockingQueue, rejectedExecutionHandler);

	while(System.currentTimeMillis() - startTime < TIME_LIMIT) {
		executor.submit(new Runnable() {
			@Override
			public void run() {
				mAuthContext.acquireTokenSilentAsync(RESOURCE_ID, CLIENT_ID, User_UPN, new AuthenticationCallback<AuthenticationResult>() {

					@Override
					public void onSuccess(AuthenticationResult authenticationResult) {}

					@Override
					public void onError(Exception exc) {}
				});
			}
		});
		Thread.sleep(5);
	}
	executor.shutdownNow();
	System.err.println( "All done." );
}
```

*Logs*
[adal_outofmemory_log.txt](https://github.com/AzureAD/azure-activedirectory-library-for-android/files/965281/adal_outofmemory_log.txt)

**Comment by heidijinxujia** [comment]
[issue824-log.txt](https://github.com/AzureAD/azure-activedirectory-library-for-android/files/963162/issue824-log.txt)

**Comment by heidijinxujia** [comment]
acquireTokenSilient async call with RT in cache also got the outOfMemory error after few mins of running.

Attach the logs here.
[adal_outofmemory_rt_log.txt](https://github.com/AzureAD/azure-activedirectory-library-for-android/files/965513/adal_outofmemory_rt_log.txt)

**Comment by weijjia** [comment]
@heidijinxujia Can you run the same profile tool as you have done for msal? I'm guess we have array holding large chunk of contiguous memory and that cause the bytes allocation of string creation fail.

**Comment by nazukj** [comment]
@heidijinxujia please prioritize this after doze mode fix. Thanks!

**Comment by heidijinxujia** [comment]
OK. Will do.

**Comment by heidijinxujia** [comment]
I will apply Ping's change and do the stress test again after finishing the doze mode testing. My test device is currently occupied.

**Comment by piqiums** [comment]
when OOM happens, the stacktrace won't help, since there's very little available memory, it could crash randomly on any scene, even requesting a buffer of several KB (like triggering a HTTP request) will cause OOM.   In this case, the .hprof file is the key to find what objects were leaked. In Android Monitor, you could 'dump Java heap'. That's a snapshot of memory in Java heap. Then with MAT, you could dive deeper into the issue.

**Comment by heidijinxujia** [comment]
Fixed in #906
